File: frontend/src/tests/unit/test_youtube_duration_integrity.mjs
// frontend/src/tests/unit/test_youtube_duration_integrity.mjs

import test from 'node:test'
import assert from 'node:assert/strict'

import {
  formatVideoDuration,
  resolveSkillVideo,
  CURATED_FRONTEND_FALLBACKS
} from '../../utils/videoResolver.js'

import {
  formatStageVideos,
  clearStageVideoCache,
  fetchStageVideos,
  setApiClient
} from '../../services/youtubeVideoService.js'

test('PHASE 3.2 — YouTube Duration & Metadata Integrity Regression Suite', async (t) => {

  // =========================================================================
  // TEST 1: W6NZfCO5SIk has the canonical duration (2896 seconds / 48:16)
  // =========================================================================
  await t.test('TEST 1: W6NZfCO5SIk has canonical duration 48:16 and 2896 seconds', () => {
    // 1. Check curated fallback catalog
    const jsFallback = CURATED_FRONTEND_FALLBACKS['javascript']
    assert.equal(jsFallback.id, 'W6NZfCO5SIk')
    assert.equal(jsFallback.channel, 'Programming with Mosh')
    assert.equal(jsFallback.duration, '48:16')
    assert.equal(jsFallback.duration_seconds, 2896)

    // 2. Check formatStageVideos normalization
    const stageVids = formatStageVideos([
      {
        id: 'W6NZfCO5SIk',
        title: 'JavaScript Tutorial for Beginners',
        channel: 'Programming with Mosh',
        url: 'https://www.youtube.com/watch?v=W6NZfCO5SIk',
        embed_url: 'https://www.youtube.com/embed/W6NZfCO5SIk',
        duration: '48:16',
        duration_seconds: 2896
      }
    ], 'JavaScript')

    assert.equal(stageVids.length, 1)
    assert.equal(stageVids[0].id, 'W6NZfCO5SIk')
    assert.equal(stageVids[0].duration, '48:16')
    assert.equal(stageVids[0].duration_seconds, 2896)
    assert.equal(stageVids[0].channel, 'Programming with Mosh')
  })

  // =========================================================================
  // TEST 2: Incorrect duration 68 minutes / "1 hr 8 mins" must never be returned
  // =========================================================================
  await t.test('TEST 2: Incorrect duration 68 minutes / "1 hr 8 mins" is never returned or displayed', () => {
    // Simulated bad payload with old '1 hr 8 mins'
    const badInput = [
      {
        id: 'W6NZfCO5SIk',
        title: 'JavaScript Tutorial',
        duration: '1 hr 8 mins',
        duration_seconds: 4080,
        url: 'https://www.youtube.com/watch?v=W6NZfCO5SIk'
      }
    ]

    const formatted = formatStageVideos(badInput, 'JavaScript')
    assert.equal(formatted.length, 1)
    // Canonical defense enforces 48:16 and 2896
    assert.equal(formatted[0].duration, '48:16')
    assert.equal(formatted[0].duration_seconds, 2896)
    assert.notEqual(formatted[0].duration, '1 hr 8 mins')

    // formatVideoDuration must also reject "1 hr 8 mins"
    assert.equal(formatVideoDuration('1 hr 8 mins', null), null)
    assert.equal(formatVideoDuration('68 mins', null), null)
  })

  // =========================================================================
  // TEST 3: Frontend correctly formats 2896 seconds to "48:16"
  // =========================================================================
  await t.test('TEST 3: Frontend duration formatter correctly handles seconds', () => {
    // 2896 seconds -> 48:16
    assert.equal(formatVideoDuration(null, 2896), '48:16')
    assert.equal(formatVideoDuration('48:16', 2896), '48:16')

    // String seconds "2896"
    assert.equal(formatVideoDuration('2896', null), '48:16')

    // Multi-hour videos
    // 4803 seconds -> 1:20:03
    assert.equal(formatVideoDuration(null, 4803), '1:20:03')
    // 15638 seconds -> 4:20:38
    assert.equal(formatVideoDuration(null, 15638), '4:20:38')

    // Sub-hour videos
    // 699 seconds -> 11:39
    assert.equal(formatVideoDuration(null, 699), '11:39')
    // 1798 seconds -> 29:58
    assert.equal(formatVideoDuration(null, 1798), '29:58')
  })

  // =========================================================================
  // TEST 4: Unknown duration does not generate a fake duration (returns null)
  // =========================================================================
  await t.test('TEST 4: Unknown duration returns null (Option A: badge hidden)', () => {
    // Null/undefined values
    assert.equal(formatVideoDuration(null, null), null)
    assert.equal(formatVideoDuration(undefined, undefined), null)
    assert.equal(formatVideoDuration('', null), null)

    // Fabricated strings that must NOT be displayed
    assert.equal(formatVideoDuration('45 mins', null), null)
    assert.equal(formatVideoDuration('40 mins', null), null)
    assert.equal(formatVideoDuration('1 hr', null), null)
    assert.equal(formatVideoDuration('Duration unavailable', null), null)

    // In formatStageVideos, unknown video without duration has null duration
    const unknownVids = formatStageVideos([
      {
        id: 'dQw4w9WgXcQ',
        title: 'Unknown Tech Topic',
        url: 'https://www.youtube.com/watch?v=dQw4w9WgXcQ'
      }
    ], 'Unknown')

    assert.equal(unknownVids.length, 1)
    assert.equal(unknownVids[0].duration, null)
    assert.equal(unknownVids[0].duration_seconds, null)
  })

  // =========================================================================
  // TEST 5: Backend and frontend use the same duration representation
  // =========================================================================
  await t.test('TEST 5: Backend and frontend use identical canonical duration for W6NZfCO5SIk', () => {
    const backendSimulated = {
      id: 'W6NZfCO5SIk',
      title: 'JavaScript Tutorial for Beginners',
      channel: 'Programming with Mosh',
      url: 'https://www.youtube.com/watch?v=W6NZfCO5SIk',
      embed_url: 'https://www.youtube.com/embed/W6NZfCO5SIk',
      duration: '48:16',
      duration_seconds: 2896
    }

    const resolved = resolveSkillVideo({
      skill_name: 'JavaScript',
      youtube_videos: [backendSimulated]
    })

    assert.equal(resolved.id, 'W6NZfCO5SIk')
    assert.equal(resolved.duration, '48:16')
    assert.equal(resolved.duration_seconds, 2896)
    assert.equal(resolved.channel, 'Programming with Mosh')

    // Formatted UI string matches
    const uiDuration = formatVideoDuration(resolved.duration, resolved.duration_seconds)
    assert.equal(uiDuration, '48:16')
  })

  // =========================================================================
  // TEST 6: Metadata integrity test
  // =========================================================================
  await t.test('TEST 6: Metadata integrity across id, title, channel, thumbnail, URL, duration', () => {
    const stageVids = formatStageVideos([
      {
        id: 'W6NZfCO5SIk',
        title: 'JavaScript Tutorial for Beginners',
        channel: 'Programming with Mosh',
        duration: '48:16',
        duration_seconds: 2896,
        url: 'https://www.youtube.com/watch?v=W6NZfCO5SIk',
        embed_url: 'https://www.youtube.com/embed/W6NZfCO5SIk',
        thumbnail: 'https://i.ytimg.com/vi/W6NZfCO5SIk/mqdefault.jpg'
      }
    ], 'JavaScript')

    const video = stageVids[0]
    assert.equal(video.id, 'W6NZfCO5SIk')
    assert.equal(video.channel, 'Programming with Mosh')
    assert.equal(video.duration, '48:16')
    assert.equal(video.duration_seconds, 2896)
    assert.ok(video.url.includes('W6NZfCO5SIk'))
    assert.ok(video.embed_url.includes('W6NZfCO5SIk'))
    assert.ok(video.thumbnail.includes('W6NZfCO5SIk'))
  })

  // =========================================================================
  // TEST 7: Duplicate video records must not have conflicting durations
  // =========================================================================
  await t.test('TEST 7: Duplicate video records deduplicate and enforce canonical duration', () => {
    const raw = [
      {
        id: 'W6NZfCO5SIk',
        title: 'JS Tutorial 1',
        duration: '1 hr 8 mins',
        url: 'https://www.youtube.com/watch?v=W6NZfCO5SIk'
      },
      {
        id: 'W6NZfCO5SIk',
        title: 'JS Tutorial 2',
        duration: '30 mins',
        url: 'https://www.youtube.com/watch?v=W6NZfCO5SIk'
      }
    ]

    const deduped = formatStageVideos(raw, 'JavaScript')
    assert.equal(deduped.length, 1)
    assert.equal(deduped[0].id, 'W6NZfCO5SIk')
    assert.equal(deduped[0].duration, '48:16')
    assert.equal(deduped[0].duration_seconds, 2896)
  })

  // =========================================================================
  // TEST 8: Cached metadata must pass duration validation
  // =========================================================================
  await t.test('TEST 8: fetchStageVideos returns verified duration and updates cache', async () => {
    clearStageVideoCache()

    // Mock API client returning W6NZfCO5SIk
    setApiClient({
      get: async () => ({
        data: {
          videos: [
            {
              id: 'W6NZfCO5SIk',
              title: 'JavaScript Tutorial for Beginners',
              channel: 'Programming with Mosh',
              duration: '48:16',
              duration_seconds: 2896,
              url: 'https://www.youtube.com/watch?v=W6NZfCO5SIk',
              embed_url: 'https://www.youtube.com/embed/W6NZfCO5SIk'
            }
          ]
        }
      })
    })

    const fetched = await fetchStageVideos({
      skill: 'JavaScript',
      targetRole: 'Software Engineer',
      stage: 'learn',
      language: 'en'
    })

    assert.equal(fetched.length, 1)
    assert.equal(fetched[0].id, 'W6NZfCO5SIk')
    assert.equal(fetched[0].duration, '48:16')
    assert.equal(fetched[0].duration_seconds, 2896)

    // Second call reads from cache
    const fromCache = await fetchStageVideos({
      skill: 'JavaScript',
      targetRole: 'Software Engineer',
      stage: 'learn',
      language: 'en'
    })

    assert.equal(fromCache.length, 1)
    assert.equal(fromCache[0].id, 'W6NZfCO5SIk')
    assert.equal(fromCache[0].duration, '48:16')
    assert.equal(fromCache[0].duration_seconds, 2896)
  })
})

File: backend/tests/unit/test_youtube_duration_integrity.py
# backend/tests/unit/test_youtube_duration_integrity.py

import pytest
from app.services.youtube_service import (
    YouTubeService,
    deduplicate_youtube_videos,
    apply_canonical_metadata,
    VERIFIED_CANONICAL_METADATA,
    SKILL_VIDEO_MAP,
    STAGE_CURATED_CATALOG
)


@pytest.fixture
def yt_service():
    yt = YouTubeService()
    yt.api_key = None  # Force curated/fallback path
    yt.clear_cache()
    return yt


# =====================================================================
# TEST 1: W6NZfCO5SIk has the correct duration (2896 seconds / 48:16)
# =====================================================================
def test_case_1_w6nzfco5sik_has_correct_duration(yt_service):
    """
    Test 1: Video W6NZfCO5SIk must return canonical duration 48:16 and 2896 seconds.
    """
    videos = yt_service.get_videos_for_skill("JavaScript", stage="learn")
    assert len(videos) > 0

    mosh_video = next((v for v in videos if v['id'] == 'W6NZfCO5SIk'), None)
    assert mosh_video is not None, "W6NZfCO5SIk should be returned for JavaScript learn stage"
    assert mosh_video['duration'] == '48:16'
    assert mosh_video['duration_seconds'] == 2896
    assert mosh_video['channel'] == 'Programming with Mosh'


# =====================================================================
# TEST 2: Incorrect duration '1 hr 8 mins' / 68 minutes must never be returned
# =====================================================================
def test_case_2_no_incorrect_68_minutes(yt_service):
    """
    Test 2: W6NZfCO5SIk must NEVER have '1 hr 8 mins', '68 mins', or ~4080 seconds.
    """
    for stage in ['learn', 'practice', 'build', 'assess']:
        videos = yt_service.get_videos_for_skill("JavaScript", stage=stage)
        for v in videos:
            if v['id'] == 'W6NZfCO5SIk':
                assert v['duration'] != '1 hr 8 mins', "Fabricated duration '1 hr 8 mins' returned!"
                assert v['duration'] != '68 mins'
                assert v.get('duration_seconds') != 4080
                assert v['duration'] == '48:16'
                assert v['duration_seconds'] == 2896


# =====================================================================
# TEST 3: Metadata integrity test
# video ID, title, channel, thumbnail, URL, duration belong to same record
# =====================================================================
def test_case_3_metadata_integrity(yt_service):
    """
    Test 3: Metadata integrity across all returned videos.
    Every video record must have consistent ID, matching thumbnail URL, matching watch URL,
    matching embed URL, and valid duration representation.
    """
    sample_skills = ['JavaScript', 'Python', 'React', 'SQL', 'MongoDB', 'Power BI', 'Terraform', 'Solidity']
    for sk in sample_skills:
        videos = yt_service.get_videos_for_skill(sk, stage='learn')
        for v in videos:
            vid_id = v['id']
            assert len(vid_id) == 11
            assert vid_id in v['url']
            assert vid_id in v['embed_url']
            assert vid_id in v['thumbnail']
            assert v['title'] and len(v['title']) > 0
            assert v['channel'] and len(v['channel']) > 0

            # If video is canonical, duration and seconds must match canonical mapping
            if vid_id in VERIFIED_CANONICAL_METADATA:
                expected = VERIFIED_CANONICAL_METADATA[vid_id]
                assert v['duration'] == expected['duration']
                assert v['duration_seconds'] == expected['duration_seconds']


# =====================================================================
# TEST 4: Unknown duration does not generate a fake duration
# =====================================================================
def test_case_4_unknown_duration_no_fake(yt_service):
    """
    Test 4: If duration cannot be verified, it must be None, NOT an arbitrary string
    like '45 mins', '40 mins', or '1 hr'.
    """
    unknown_item = {
        'id': 'unknown9999',
        'title': 'Obscure Framework Guide',
        'channel': 'Random Channel',
        'duration': '45 mins'  # Fabricated duration passed into pipeline
    }
    sanitized = apply_canonical_metadata(unknown_item)
    # Fabricated duration must be stripped to None
    assert sanitized['duration'] is None
    assert sanitized.get('duration_seconds') is None

    # Test with empty/None
    unknown_item2 = {
        'id': 'unknown8888',
        'title': 'Another Guide'
    }
    sanitized2 = apply_canonical_metadata(unknown_item2)
    assert sanitized2.get('duration') is None
    assert sanitized2.get('duration_seconds') is None


# =====================================================================
# TEST 5: Duplicate video records must not have conflicting durations
# =====================================================================
def test_case_5_no_conflicting_durations_in_duplicates():
    """
    Test 5: Duplicate occurrences of the same video ID in raw input must resolve
    to the single canonical duration without conflicting values.
    """
    raw = [
        {'id': 'W6NZfCO5SIk', 'title': 'JS 1', 'duration': '1 hr 8 mins'},
        {'id': 'W6NZfCO5SIk', 'title': 'JS 2', 'duration': '30 mins'},
        {'id': 'ofme2o29ngU', 'title': 'Mongo 1', 'duration': '45 mins'},
        {'id': 'ofme2o29ngU', 'title': 'Mongo 2', 'duration': '29:58'}
    ]
    cleaned = deduplicate_youtube_videos(raw)
    assert len(cleaned) == 2
    assert cleaned[0]['id'] == 'W6NZfCO5SIk'
    assert cleaned[0]['duration'] == '48:16'
    assert cleaned[0]['duration_seconds'] == 2896

    assert cleaned[1]['id'] == 'ofme2o29ngU'
    assert cleaned[1]['duration'] == '29:58'
    assert cleaned[1]['duration_seconds'] == 1798


# =====================================================================
# TEST 6: Cached metadata must pass duration validation
# =====================================================================
def test_case_6_cached_metadata_duration_validation(yt_service):
    """
    Test 6: In-memory cache must return verified canonical duration even if
    stale data was initially populated.
    """
    cache_key = "javascript|software engineer|learn|en|4"
    # Inject a simulated stale cache entry with '1 hr 8 mins'
    yt_service._cache[cache_key] = [
        {'id': 'W6NZfCO5SIk', 'title': 'JavaScript Tutorial for Beginners', 'duration': '1 hr 8 mins'}
    ]

    # Calling get_videos_for_skill will access the cache
    videos = yt_service.get_videos_for_skill("JavaScript", stage="learn")
    assert len(videos) == 1
    assert videos[0]['id'] == 'W6NZfCO5SIk'
    # The cache read applies canonical metadata and sanitizes the duration
    assert videos[0]['duration'] == '48:16'
    assert videos[0]['duration_seconds'] == 2896


# =====================================================================
# TEST 7: Multi-video canonical duration audit across skills
# =====================================================================
def test_case_7_multi_video_duration_audit(yt_service):
    """
    Test 7: Audits multiple curated videos across skills to verify their
    exact YouTube duration and seconds are preserved.
    """
    test_cases = [
        ('Python', 'learn', 'rfscVS0vtbw', '4:26:51', 16011),
        ('SQL', 'learn', 'HXV3zeQKqGY', '4:20:38', 15638),
        ('React', 'learn', 'bMknfKXIFA8', '11:55:27', 42927),
        ('MongoDB', 'learn', 'ofme2o29ngU', '29:58', 1798),
        ('Power BI', 'learn', 'TmhQCQr_DCA', '27:52', 1672),
        ('Terraform', 'learn', '7xngnjfIlK4', '2:38:03', 9483),
        ('Solidity', 'learn', 'M576WGiDBdQ', '16:22:11', 58931)
    ]

    for skill, stage, expected_id, expected_dur, expected_sec in test_cases:
        videos = yt_service.get_videos_for_skill(skill, stage=stage)
        match = next((v for v in videos if v['id'] == expected_id), None)
        assert match is not None, f"Video {expected_id} not found for {skill} ({stage})"
        assert match['duration'] == expected_dur, f"Mismatch for {skill}: expected {expected_dur}, got {match['duration']}"
        assert match['duration_seconds'] == expected_sec, f"Seconds mismatch for {skill}: expected {expected_sec}, got {match['duration_seconds']}"
